File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/base/fragment/BaseDialogFragment.java
package ua.meugen.android.pdfcomposer.ui.activities.base.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;

import ua.meugen.android.pdfcomposer.app.PdfComposer;
import ua.meugen.android.pdfcomposer.ui.activities.base.Injector;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter.MvpPresenter;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.BaseState;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.MvpState;
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.view.MvpView;


public abstract class BaseDialogFragment<S extends MvpState, P extends MvpPresenter<S>>
        extends DialogFragment implements MvpView {

    public P presenter;
    public S state;

    @Override
    public void onAttach(final Context context) {
        inject(context);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            ((BaseState) state).attachBundle(getArguments());
        } else {
            ((BaseState) state).attachBundle(savedInstanceState);
        }
        presenter.onRestoreState(state);
        ((BaseState) state).detachBundle();
    }

    @Override
    public void onSaveInstanceState(final Bundle outState) {
        super.onSaveInstanceState(outState);
        ((BaseState) state).attachBundle(outState);
        presenter.onSaveState(state);
        ((BaseState) state).detachBundle();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenter.clear();
    }

    private void inject(final Context context) {
        final Injector injector = createInjector();
        injector.inject(PdfComposer.from(context).getAppComponent());
    }

    protected abstract Injector createInjector();
}

File: app/src/main/java/ua/meugen/android/pdfcomposer/ui/activities/base/fragment/BaseDialogFragment.kt
package ua.meugen.android.pdfcomposer.ui.activities.base.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.DialogFragment

import ua.meugen.android.pdfcomposer.app.PdfComposer
import ua.meugen.android.pdfcomposer.ui.activities.base.Injector
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.presenter.MvpPresenter
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.BaseState
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.state.MvpState
import ua.meugen.android.pdfcomposer.ui.activities.base.fragment.view.MvpView


abstract class BaseDialogFragment<S : MvpState, P : MvpPresenter<S>> : DialogFragment(), MvpView {

    lateinit var presenter: P
    lateinit var state: S

    override fun onAttach(context: Context?) {
        inject(context)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState == null) {
            (state as BaseState).attachBundle(arguments)
        } else {
            (state as BaseState).attachBundle(savedInstanceState)
        }
        presenter.onRestoreState(state)
        (state as BaseState).detachBundle()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        (state as BaseState).attachBundle(outState)
        presenter.onSaveState(state)
        (state as BaseState).detachBundle()
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.clear()
    }

    private fun inject(context: Context?) {
        val injector = createInjector()
        injector.inject(PdfComposer.from(context!!).appComponent)
    }

    protected abstract fun createInjector(): Injector
}
